doc/ref.man
.TH REF 1
.SH NAME
ref \- Display a C function header
.SH SYNOPSIS
.B ref
.RI [ options ]...
.RI [ restrictions ]...
.SH VERSION
This page describes the
.B Elvis 2.2_1
version of
.BR ref .
See
.BR elvis (1).
.SH DESCRIPTION
.B ref
quickly locates and displays the header of a function.
To do this,
.B ref
looks in the "tags" file for the line that describes the function, and then 
scans the source file for the function.
When it locates the function, it displays an introductory comment
(if there is one), the function's declaration, and the declarations of all
arguments.
.SH OPTIONS
.IP \fB\-t\fR
Output tag info, instead of the function header.
The tag info consists of the three standard fields from each tag.
The fields will be separated by tab characters, just like records from the
traditional tags file format.
Consequently, you can use "ref \-t \-a >oldtags" to convert a new-style tags file
back to an old-style tags file.
.IP \fB\-v\fR
Output verbose tag info, instead of the function header.
The verbose tag info shows the names and values of all attributes for each
matching tag.
Each name/value pair is shown on a separate line.
It also shows the "match" factor that is used for sorting tags which have
the same tagname.
.IP \fB\-h\fR
Output HTML browser info, instead of the function header.
This is an HTML table with hypertext links into the source files where the
tags are defined.
You can use Netscape or another browser to use this, but they won't move
the cursor to the correct line within the source file; only
.B Elvis
knows how to do that.
.IP
This resembles
.BR Elvis '
.B :browse
command.
.IP \fB\-c\fR
Don't output introductory comments before each tag definition line.
.IP \fB\-d\fR
Don't output other lines of the definition.
The line where the tag is defined is shown but any preceding or following
lines which are part of the definition will be omitted.
.IP \fB\-a\fR
List all matching tags.
(Without this option,
.B ref
would only output the single most likely tag.)
.IP \fB\-s\fR
Search all tags files.
Without this option,
.B ref
would stop searching after processing the first tags file which 
contained any tags which met the restrictions.
.IP "\fB\-p\fI tagpath\fR"
List of directories or tags files to search through.
By default,
.B ref
would use the value from the TAGPATH environment variable or a hardcoded
default value for each operating system.
.IP "\fB\-l\fI taglength\fR"
Only check the first
.I taglength
characters of tag names.
The default behavior is to check all characters.
.SH "RESTRICTIONS AND HINTS"
Other than options, any argument on
.BR ref 's
command line is interpreted as a restriction or sorting hint.
Elvis parses all of the restrictions and sorting hints, and then scans
the tags files (i.e., every file listed in the tag path, or a file named
"tags" in every directory listed in the tag path).
Tags which fail to meet any restriction are discarded.
.PP
Other tags are inserted into a list.
The list is sorted primarily by each tag's tagname.
If multiple tags have the same overloaded name, then those tags will be
sorted according to the sorting hints.
In the absence of hints, the tags will be added in the same order in which
they appear in the tags file.
.PP
The restrictions can be given in any of the following forms:
.TP
.IB name : value [, value\fB...]
Reject tags which have an attribute named
.IR name ,
but that attribute's value isn't in the list of acceptable values.
E.g., "class:Foo" rejects tags from a different class,
but accepts tags which have no class.
.TP
.IB name := value [, value\fB...]
Reject tags which have an attribute named
.IR name ,
but that attribute's value isn't in the list of acceptable values.
Also reject tags which don't have an attribute named
.IR name .
E.g., "class:=Foo" only accepts tags which have class "Foo".
.TP
.IB name :/ value [, value\fB...]
Like "name:value" except that the
.B tagaddress
field is required to contain
.I value
as a substring.
So "class:/Foo" would find tags in class "Foo" PLUS
global tags whose address mentions "Foo" -
probably friends of the Foo class.
.TP
.IB value [, value ...]
Short for 
.BI tagname: value [, value ...]
.PP
The sorting hints follow a similar form:
.TP
.IB name :+ value [, value\fB...]
Pretend that recent successful searches had attributes named "name" with the
given values.
This causes any similar tags in the new search to appear near the top of the
list.
.TP
.IB name :- value [, value\fB...]
Pretend that recent failed searches had attribute named "name" with the
given values.
This causes any similar tags in the new search to appear near the bottom of the
list.
.PP
A null value string matches anything.
So "struct:=" would accept any tag with a "struct" attribute, and
reject those without it.
This would be handy when you're trying to do tag lookup for a word which
follows a '.' character -
you know it is a field name, but you don't know from which struct type.
.PP
Note that if you invoke
.B ref
without giving any restrictions,
then
.B all
tags will match and will (if invoked with the
.B \-a
flag) be output.
.SH "A REAL-WORLD EXAMPLE"
While converting some code from K&R C to ANSI C,
I needed to generate extern declarations for all the functions.
To find the global function headers, I used the command...
.nf

		ref \-a kind:f file:dummy
.fi
.PP
The "\-a" causes
.B ref
to output all headers, instead of just the first one that it finds.
"kind:f" causes it to exclude any non-functions.
"file:dummy" is tricky -- it causes
.B ref
to exclude static tags from all files except "dummy", and since there were
no C functions defined in any file named "dummy", all statics were excluded.
I only got globals.
.PP
Once I had a list of all global functions, I still had to do some editing
to convert them into ANSI declarations
.RB ( ref
couldn't help me there)
but at least I could be confident
that my list of functions was complete and accurate.
.PP
For each source file, I also needed to find the static functions defined
there, so for each "file.c" I used the command...
.nf

		ref \-a kind:f file:=file.c
.fi
.PP
This is very similar to the earlier command.
The main difference is that we're using "file:="
(with an '=', to exclude globals)
and a real file name (instead of "dummy") so we do include the static
tags from that particular file.

.SH "INTERACTION WITH ELVIS"
.B ref
is used by
.BR elvis (1)'s
shift-K command.
If the cursor is located on a word such as "splat", in the file "foo.c",
then
.B Elvis
will invoke
.B ref
with the command "ref splat file:foo.c".
.PP
Elvis' "smartargs" option also uses
.B ref
to look up the arguments for a given function.

.SH TAGS
A tag is a collection of attributes.
Each attribute has a name and a value.
Every tag has attributes with the following names:
.TP
.B tagname
The name of the tag; usually the same as the function (or whatever) that the
tag refers to.
.TP
.B tagfile
The name of your source code file, in which the tag's definition occurred.
.TP
.B tagaddress
Either a line number, or a "nomagic" regular expression, which allows
.B Elvis
or
.B ref
to locate the tag's definition within your source file.
.PP
In addition, any tag can have additional, optional attributes.
These extra tags are meant to serve as hints, describing the contexts in which
the tagname is permitted to occur in your source code.
The list of additional attribute names is not preset; any tags file can
use whatever seem appropriate.
The following are typical:
.TP
.B kind
This value is a single letter indicating the lexical type of the tag.
It can be "f" for functions, "v" for variables, and so on.
.TP
.B file
If the tag can only be used within a single source file, then this should be
the name of that file.
E.g., in C, a "static" function can only be used in the function in which it
is defined, so if a function is static then its tag will usually have a
.B file
attribute, and its value will be the same as that of its
.B tagfile
attribute.
.TP
.B function
For local variables.
The value is name of function in which they're defined.
.TP
.B struct
For fields of a struct or union.
The value is the name of the struct or union.
If it has no name (not even a typedef) then "struct=struct"
is better than nothing.
.TP
.B enum
For values in an enum data type.
The value is the name of the enum type.
If it has no name (not even a typedef) then "enum=enum"
is better than nothing.
.TP
.B class
Member functions of a class in C++ could use this to identify which class
they're in.
The class name itself, however, is global so it doesn't have a
.B class
attribute.
.TP
.B scope
Intended mostly for class member functions.
It will usually be "public" or "private",
so users can restrict tag searches to only public members.
.TP
.B arity
For functions.
Its value is the number of arguments.
.PP
Currently, the hacked-up version of
.BR ctags (1)
(sometimes installed as
.BR elvtags (1))
included with
.B Elvis
will only generate
.BR kind ,
.BR file ,
and
.B class
hints, and it doesn't do a very good job on
.B class
hints.
.SH "THE TAGS FILE"
The tags file is a text file, in which each line describes a single tag.
Each line is divided into fields, delimited by tab characters.
.PP
The first 3 fields are implicitly defined to be the values of the
tagname, tagfile, and tagaddress attributes, in that order.
Note that this is identical to the traditional format of the tags file.
.PP
If there are other fields, then semicolon-doublequote will be appended to
the tagaddress field; vi ignores anything after that, so the extra fields
won't interfere with vi's ability to perform tag searches.  Other editors
such as
.B Elvis
and
.B Vim
use the extra fields though.
.PP
The extra fields are required to have the format "<tab>name:value".
I.e., a ':' is required, and
everything before the ':' is used as an attribute name, and
everything after it is used as this tag's value for that attribute.
There are two exceptions:
.IP * 3n
If an extra field lacks a colon, then the field is
assumed to be the value of an attribute named "kind".  (Some versions
of ctags generate a single-letter "kind" value to indicate whether a
tag is a
.BR f unction,
.BR v ariable,
.BR t ypedef,
etc.)
Since nearly all tags have this field, omitting "kind:" significantly
reduces the size of the tags file, and the time needed to search it.
.IP * 3n
Static tags are usually marked with "file:", with no file name after the ":".
In this case the file name is understood to be identical to the "tagfile" field.
This does more than just reduce the size of the tags file --
"tagfile" values are relative to the directory containing the tags file,
and this rule offers a way to make "file" values be relative, too.
.PP
Different tags may have differently named  hints.
Since each hint includes an explicit name with each value,
they can appear in any order, and
you can omit any which don't apply to a given tag.
.PP
.B Ref
and
.B Elvis
store attribute names are stored in a fixed-size array, which is
shared among all tags from a given file.
Consequently, the number of distinct attribute names within a tags file
is limited.
As currently configured, that limit is 10 names -
the 3 standard ones plus up to 7 other names for hints.
.SH "THE REFS FILE"
When
.B ref
has found a tag entry and is searching for the source of that tag,
if it can't read the original source file then it will try to read a file
named "refs".
The "refs" file should contain a copy of all source code, with the
bodies of functions replaced by "{}".
Elvis' version of
.BR ctags (1)
can generate a "refs" file.
.SH FILES
The following files can be found in any directory named in the tagpath.
.TP
.I tags
List of function names and their locations, generated by
.BR ctags (1).
.TP
.I refs
Function headers extracted from source files (optional).
.SH ENVIRONMENT
.IP \fBTAGPATH\fR
List of directories or files to be searched.
In the case of directories,
.B ref
looks for a file named "tags" in that directory.
The elements in the list are separated by either
colons (for Unix) or semicolons (for most other operating systems).
For each operating system,
.B ref
has a built-in default which is probably adequate.
.SH NOTES
You might want to generate a "tags" file for the directory that contains the
source code for standard C library on your system.
This will allow
.B ref
to serve as a quick reference for any library function in addition to your
project's functions.
.PP
If licensing restrictions prevent you from making the library source readable
by everybody, then you can have
.BR Elvis '
version of
.BR ctags (1)
(often installed as "elvtags")
generate a "refs" file, and make "refs" readable by everybody.
If your system doesn't come with the library source code, then perhaps you
can produce something workable from the
.BR lint (1)
libraries.
.SH "SEE ALSO"
.BR elvis (1),
.BR ctags (1),
.BR lint (1)
.P
Note that on most Unix systems, the
.B Elvis
version of
.BR ctags (1)
is installed as
.BR elvtags (1).

.SH AUTHOR
Steve Kirkendall
.br
kirkenda@cs.pdx.edu


debian/NEWS
elvis (2.2.0-2) unstable; urgency=low

  The elvis package has been split into elvis-common, elvis,
  elvis-console and elvis-tools.

    elvis-common:
      This package provides the common files (manual, ex scripts,
      digraph files, etc.) that are needed by the packages elvis,
      elvis-console and elvis-tools.

    elvis:
      This package contains Elvis compiled with support for X11, Xft
      and background images.  If you don't need the X11 interface,
      install elvis-console instead.

    elvis-console:
      This package contains Elvis compiled without X11 support.  If
      you want to use Elvis' X11 interface then install the package
      elvis instead.

    elvis-tools:
      This package contains elvtags (for generating tags and refs
      files), ref (for quickly locating/displaying the header of a
      function) and elvfmt (for adjusting paragraphs of text).
      These tools actually belong to the Elvis editor but can be used
      without it.

  As a consequence of this, elvfmt, elvref and elvtags will be removed
  if you're upgrading from an earlier version of the elvis package.
  In this case, just install elvis-tools afterwards if you need them.
  Also note that elvref is now again installed as ref.

 -- Georg Neis <georg@f-451.net>  Tue, 04 Aug 2004 13:12:55 +0200


ref.c
/* ref.c */

/* This program locates a tag, and displays its context.  A handy reference. */

#include "elvis.h"

/* This is the default path that is searched for tags */
#if OSK
# define DEFTAGPATH ".:/dd/defs:/dd/defs/sys:/dd/usr/src/lib:../lib:/dd/usr/lib"
#else
# if ANY_UNIX
#  define DEFTAGPATH ".:/usr/include:/usr/include/sys:/usr/src/lib:../lib:/usr/local/lib"
# else
#  if MSDOS || TOS || OS2
#   define DEFTAGPATH ".;C:\\include;C:\\include\\sys;C:\\lib;..\\lib"
#  else
#   if AMIGA
#    define DEFTAGPATH ".;Include:;Include:sys"
#   else /* any other OS */
#    define DEFTAGPATH "."
#   endif
#  endif
# endif
#endif

/* maximum length of a path list */
#define MAXPATH	2048

/* compile some of the osdir functions into this module */
#define JUST_DIRPATH
#include "osdir.c"

/* This data type is used for classifying lines */
typedef enum
{
	LC_BLANK,	/* an empty line */
	LC_COMMENT,	/* a comment, or continuation of a comment */
	LC_PARTIAL,	/* partial definition -- no semicolon, or in brackets */
	LC_COMPLETE	/* anything else */
} LINECLS;


#if USE_PROTOTYPES
static void usage(char *argv0);
static char *elv_getline(FILE *fp);
static void store(char *line, char **list);
static LINECLS classify(char *line, LINECLS prev);
static void lookup(TAG *tag);
static void add_to_path(char *file);
int main(int argc, char **argv);
#endif /* USE_PROTOTYPES */


/* support for elvis' ctype macros */
#ifdef ELVCT_DIGIT
CHAR elvct_class[256] = {
	0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
	0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
	ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,
	ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,ELVCT_DIGIT,
	0,0,0,0,0,0,0,
	ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,
	ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,
	ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,
	ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,
	ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,ELVCT_UPPER,
	ELVCT_UPPER,
	0,0,0,0,0,0,
	ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,
	ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,
	ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,
	ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,
	ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,ELVCT_LOWER,
	ELVCT_LOWER
};
#endif


/* These reflect the command-line options */
static int output_tag_info;
static int output_verbose_info;
static int output_html_browser;
static int output_all_matches;
static int omit_comment_lines;
static int omit_other_lines;
static int search_all_files;
static char tag_path[MAXPATH] = DEFTAGPATH;
static long tag_length;


/* These store lines which preceed the tag definition. */
static char	*comments[20];	/* comment lines before tag */
static int	ncomments;	/* number of comment lines */
static char	*members[20];	/* partial definitions before tag */
static int	nmembers;	/* number of partial definition lines */

static void usage(argv0)
	char *argv0;	/* name of program */
{
	fprintf(stderr, "Usage: %s [options] [restrictions]...\n", argv0);
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "   -t             Output tag info, instead of the function header\n");
	fprintf(stderr, "   -v             Output verbose tag info, instead of the function header\n");
	fprintf(stderr, "   -h             Output HTML browser info, instead of the function header\n");
	fprintf(stderr, "   -c             Don't output introductory comments\n");
	fprintf(stderr, "   -d             Don't output other lines of the definition\n");
	fprintf(stderr, "   -a             List all matches (else just the most likely one)\n");
	fprintf(stderr, "   -s             Search all tags files (else stop after first with matches)\n");
	fprintf(stderr, "   -p tagpath     List of directories or tags files to search\n");
	fprintf(stderr, "   -l taglength   Only check the first 'taglength' characters of tag names\n");
	fprintf(stderr, "Restrictions:\n");
	fprintf(stderr, "   tag            A tag to search for, short for tagname:tag\n");
	fprintf(stderr, "   attrib:value   An optional attribute (global tags permitted)\n");
	fprintf(stderr, "   attrib:=value  A mandatory attribute (global tags rejected)\n");
	fprintf(stderr, "   attrib:/value  An optional attribute, but require as substring of tagaddr\n");
	fprintf(stderr, "   attrib:+value  List tags with given attribute first (more likely)\n");
	fprintf(stderr, "   attrib:-value  List tags with given attribute last (less likely)\n");
	fprintf(stderr, "A single attribute can be given multiple acceptable values.  The easiest way\n");
	fprintf(stderr, "to specify them is \"attrib:value,value,value\".  This works for any of the\n");
	fprintf(stderr, "restriction operators.  A global tag is one which has no value for a given\n");
	fprintf(stderr, "attribute; e.g., for the \"class\" attribute, a tag with no class is global.\n");
	exit(0);
}


/* display a fatal error message from safe.c */
#if USE_PROTOTYPES
void msg(MSGIMP type, char *msg, ...)
#else
void msg(type, msg)
	MSGIMP	type;
	char	*msg;
#endif
{
	fprintf(stderr, "%s\n", msg);
	abort();
}


/* some custom versions of elvis text I/O functions */
static FILE *iofp;
#ifdef DEBUG_ALLOC
ELVBOOL _ioopen(file, line, name, rwa, prgsafe, force, eol)
	char	*file;
	int	line;
#else
ELVBOOL ioopen(name, rwa, prgsafe, force, eol)
#endif
	char	*name;	/* name of file to open */
	_char_	rwa;	/* ignored; 'r'=read, 'w'=write, 'a'=append */
	ELVBOOL	prgsafe;/* ignored; safe to use "!prg" as file name? */
	ELVBOOL	force;	/* ignored; okay to overwrite existing files? */
	_char_	eol;	/* ignored; open in binary mode? */
{
	iofp = fopen(name, "r");
	return (ELVBOOL)(iofp != NULL);
}
int ioread(iobuf, len)
	CHAR	*iobuf;	/* Input buffer */
	int	len;	/* maximum number of CHARs to read into iobuf */
{
	return fread(iobuf, sizeof(CHAR), len, iofp);
}
ELVBOOL ioclose()
{
	fclose(iofp);
	return ElvTrue;
}


/* This function reads a single line, and replaces the terminating newline with
 * a '\0' byte.  The string will be in a static buffer.  Returns NULL at EOF.
 */
static char *elv_getline(fp)
	FILE	*fp;
{
	int	ch;
 static char	buf[1024];
 	int	len;

	/* initialize ch just to avoid a bogus compiler warning */
	ch = 0;

	for (len = 0; len < QTY(buf) - 1 && (ch = getc(fp)) != EOF && ch != '\n'; )
	{
		buf[len++] = ch;
	}
	if (ch == EOF)
		return NULL;
	buf[len] = '\0';
	if (len >= 1 && buf[len - 1] == '\r')
		buf[len - 1] = '\0';
	return buf;
}


/* Store a line in a list, or clobber the list. */
static void store(line, list)
	char	*line;	/* the text to store, or NULL to clobber */
	char	**list;	/* either comments[] or members[] */
{
	int	qty;	/* size of the array */
	int	*nptr;	/* pointer to number of items already in the list */
	int	i;	/* in case we need to scroll the list */


	/* get the list specifics */
	if (list == comments)
		qty = QTY(comments), nptr = &ncomments;
	else
		qty = QTY(members), nptr = &nmembers;

	/* if supposed to clobber, then clobber */
	if (!line)
	{
		for (i = 0; i < *nptr; i++)
		{
			safefree(list[i]);
		}
		*nptr = 0;
		return;
	}

	/* if list is full, then scroll it */
	if (*nptr == qty)
	{
		safefree(list[0]);
		for (i = 1; i < *nptr; i++)
		{
			list[i - 1] = list[i];
		}
		(*nptr)--;
	}

	/* add the new line to the list */
	list[(*nptr)++] = safedup(line);
}


/* This function classifies a line */
static LINECLS classify(line, prev)
	char	*line;	/* the line to classify */
	LINECLS prev;	/* classification of previous line */
{
	char	*front;	/* line, after skipping indentation */

	/* find the front of the line */
	for (front = line; *front == ' ' || *front == '\t'; front++)
	{
	}

	/* blank line? */
	if (!*front)
		return LC_BLANK;

	/* is it the start of a comment? */
	if ((front[0] == '/' && (front[1] == '*' || front[1] == '/'))
	 || (front[0] == '(' && front[1] == '*')
	 || (front[0] == '-' && front[1] == '-'))
	{
		return LC_COMMENT;
	}

	/* is it a continuation of a comment? */
	if (prev == LC_COMMENT
	 && !(front[0] == '#' || elvalnum(*front)))
	{
		return LC_COMMENT;
	}

	/* is it a partial declaration? */
	if (*front == '#'
		? front[strlen(front) - 1] == '\\'
		: (strchr(front, ';') == NULL
			|| (prev == LC_PARTIAL && front != line)))
	{
		return LC_PARTIAL;
	}

	/* anything else is considered to be a complete declaration */
	return LC_COMPLETE;
}

/* display the source lines which define a given tag */
static void lookup(tag)
	TAG	*tag;	/* the tag to be displayed */
{
	char	*line;	/* current line from file */
	long	lnum;	/* current line number */
	char	*l, *t;	/* for scanning chars in line and tag->TAGADDR*/
	FILE	*fp;	/* source file */
	long	taglnum;/* line number of number tag address, or 0 */
	char	*tagline;/* text form of regexp tag address */
	LINECLS	lc;	/* line classification */
	int	len;
	int	i;

	/* some initializations just to avoid bogus compiler warnings */
	tagline = NULL;
	len = 0;

	/* open the file, or the "refs" file if the source file is unreadable */
	fp = fopen(tag->TAGFILE, "r");
	if (!fp)
	{
		fp = fopen(dirpath(dirdir(tag->TAGFILE), "refs"), "r");
		if (!fp)
		{
			/* can't open anything -- give error for source file */
			(void)fopen(tag->TAGFILE, "r");
			perror(tag->TAGFILE);
			exit(1);
		}
	}

	/* initially we have no stored lines */
	store(NULL, comments);
	store(NULL, members);

	/* Convert tag address */
	taglnum = atol(tag->TAGADDR);
	if (!taglnum)
	{
		/* NOTE: We alter the regexp string!  This is okay since "ref"
		 * only searches for each tag once.
		 */
		tagline = tag->TAGADDR;
		for (l = tagline, t = tag->TAGADDR + 2; t[2]; )
		{
			if (*t == '\\')
				t++;
			*l++ = *t++;
		}
		len = (int)(l - tagline);
		if (*t == '$')
			len++;
		*l = '\0';

		/* If only supposed to output the source line (no comments or
		 * other definition lines), and the tagaddress contains the
		 * source line, then just output that.
		 */
		if (omit_comment_lines && omit_other_lines && *t == '$')
		{
			puts(tagline);
			return;
		}
	}

	/* for each line... */
	for (lnum = 1, lc = LC_COMPLETE; (line = elv_getline(fp)) != NULL; lnum++)
	{
		/* is this the tag definition? */
		if (taglnum > 0 ? taglnum == lnum : !strncmp(tagline, line, len))
		{
			/* output the tag location */
			if (!omit_comment_lines)
				printf("\"%s\", %s, line %ld:\n", tag->TAGNAME, tag->TAGFILE, lnum);

			/* output any introductory comments */
			if (!omit_comment_lines)
				for (i = 0; i < ncomments; i++)
					puts(comments[i]);

			/* output any partial definition lines */
			if (!omit_other_lines)
				for (i = 0; i < nmembers; i++)
					puts(members[i]);

			/* output this line */
			puts(line);

			/* output any following argument lines, unless the
			 * line ends with a semicolon.
			 */
			if (!omit_other_lines && line[strlen(line) - 1] != ';')
			{
				if (strchr(line, '(') != NULL)
				{
					while ((line = elv_getline(fp)) != NULL
					    && *line
					    && ((*line != '#' && *line != '{')
						|| line[strlen(line) - 1] == '\\'))
					{
						puts(line);
					}
				}
				else if ((lc = classify(line, lc)) == LC_PARTIAL)
				{
					while ((line = elv_getline(fp)) != NULL
					    && (lc = classify(line, lc)) == LC_PARTIAL)
					{
						puts(line);
					}
					if (lc == LC_COMPLETE)
						puts(line);
				}
			}

			/* done! */
			goto Succeed;
		}

		/* classify this line */
		lc = classify(line, lc);

		/* process each line, to adjust the stored lines */
		switch (lc)
		{
		  case LC_COMMENT:
			store(line, comments);
			store(NULL, members);
			break;

		  case LC_PARTIAL:
			/* leave comments unchanged, but... */
			store(line, members);
			break;

		  default:
			store(NULL, comments);
			store(NULL, members);
			break;
		}
	}

/* Fail:*/
	/* complain: not found */
	fprintf(stderr, "%s: not found in %s\n", tag->TAGNAME, tag->TAGFILE);

Succeed:
	/* close the file */
	fclose(fp);
}

/* Add the directory portion of a file name to the path */
static void add_to_path(file)
	char	*file;
{
	int	len;

	len = strlen(tag_path);
	tag_path[len++] = OSPATHDELIM;
	strcpy(tag_path + len, dirdir(file));
}

/* The main function */
int main(argc, argv)
	int	argc;
	char	**argv;
{
	int	i, j;
	char	*dir, *file, *scan;
	TAG	*tag;
	char	kindf[10];

	/* check for some standard arguments */
	if (argc > 1)
	{
		if (!strcmp(argv[1], "-help")	/* old GNU */
		 || !strcmp(argv[1], "--help")	/* new GNU */
		 || !strcmp(argv[1], "/?")	/* DOS */
		 || !strcmp(argv[1], "-?"))	/* common */
		{
			usage(argv[0]);
		}
		if (!strcmp(argv[1], "-version")  /* old GNU */
		 || !strcmp(argv[1], "--version"))/* new GNU */
		{
			printf("ref (elvis) %s\n", VERSION);
#ifdef COPY1
			puts(COPY1);
#endif
#ifdef COPY2
			puts(COPY2);
#endif
#ifdef COPY3
			puts(COPY3);
#endif
#ifdef COPY4
			puts(COPY4);
#endif
#ifdef COPY5
			puts(COPY5);
#endif
#ifdef PORTEDBY
			puts(PORTEDBY);
#endif
			exit(0);
		}
	}

	/* check the environment for TAGPATH */
	scan = getenv("TAGPATH");
	if (scan)
		strcpy(tag_path, scan);

	/* parse the options */
	for (i = 1; i < argc && argv[i][0] == '-'; i++)
	{
		for (j = 1; argv[i][j]; j++)
		{
			switch (argv[i][j])
			{
			  case 't':
				output_tag_info = 1;
				break;

			  case 'v':
				output_verbose_info = 1;
				break;

			  case 'h':
			  	output_html_browser = 1;
			  	break;

			  case 'c':
			  	omit_comment_lines = 1;
			  	break;

			  case 'd':
			  	omit_other_lines = 1;
			  	break;

			  case 'a':
				output_all_matches = 1;
				break;

			  case 's':
			  	search_all_files = 1;
			  	break;

			  case 'p':
			  	if (argv[i][j + 1])
			  		strcpy(tag_path, &argv[i][j + 1]);
			  	else if (i + 1 < argc)
					strcpy(tag_path, argv[++i]);
			  	else
			  		usage(argv[0]);
				j = strlen(argv[i]) - 1; /* skip to next argv */
				break;

			  case 'l':
			  	if (argv[i][j + 1])
			  		tag_length = atol(&argv[i][j + 1]);
			  	else if (i + 1 < argc)
			  		tag_length = atol(argv[++i]);
			  	else
			  		usage(argv[0]);
			  	if (tag_length < 0)
			  		usage(argv[0]);
				j = strlen(argv[i]) - 1; /* skip to next argv */
				break;

			  default:
			  	usage(argv[0]);
			}
		}
	}

	/* -h implies -a; nobody would want to browse a single tag */
	if (output_html_browser)
		output_all_matches = 1;
	if (output_html_browser + output_tag_info + output_verbose_info > 1)
	{
		fprintf(stderr, "%s: can't mix -t, -v, and -h\n", argv[0]);
		exit(1);
	}

	/* parse any restrictions */
	tsreset();
	strcpy(kindf, "kind:+f");
	tsparse(kindf);
	j = i;
	for ( ; i < argc; i++)
	{
		/* a little extra work for "file:" -- add its directory name
		 * to the tag path.
		 */
		if (!strncmp(argv[i], "file:", 5))
		{
			if (strchr("+-=/", argv[i][5]))
				add_to_path(&argv[i][6]);
			else
				add_to_path(&argv[i][5]);
		}

		tsparse(argv[i]);
	}

	/* As a last resort, add elvis' data directory path to the tag_path */
	i = strlen(tag_path);
	tag_path[i++] = OSPATHDELIM;
	scan = getenv("ELVISPATH");
	if (!scan)
		scan = OSLIBPATH;
	strcpy(tag_path + i, scan);

	/* for each element of the tag path... */
	for (dir = tag_path; *dir && (search_all_files || !taglist); dir = scan)
	{
		/* find the end of this directory name */
		for (scan = dir; *scan && *scan != OSPATHDELIM; scan++)
		{
		}
		if (*scan)
			*scan++ = '\0';

		/* first check to see if there is a tags file there */
		file = dirpath(*dir ? dir : ".", "tags");
		if (ioopen(file, 'r', ElvFalse, ElvFalse, 't'))
		{
			/* yes, scan the tags file */
			ioclose();
			tsfile(file, tag_length);
		}
		else
		{
			/* no, perhaps this tag element is a file? */
			tsfile(dir, tag_length);
		}
	}

	/* if nothing found, then complain */
	if (!taglist)
	{
		fprintf(stderr, "%s: tag not found\n", argv[0]);
		exit(1);
	}

	/* output HTML header, if appropriate */
	if (output_html_browser)
	{
		printf("<html><head>\n");
		printf("<title>Tag Browser</title>\n");
		printf("</head><body>\n");
		printf("<h1>Tag Browser</h1>\n");
		printf("<table border=2 cellspacing=0>\n");
	}

	/* output the information */
	for (tag = taglist; tag; tag = tag->next)
	{
		/* skip tags whose name starts with "!_".  They're pseudo-tags
		 * added to describe the tags file, not any of the user's
		 * source files.  The user doesn't care about them.
		 */
		if (!strncmp(tag->TAGNAME, "!_", 2))
			continue;

		/* output the tag */
		if (output_tag_info)
		{
			printf("%s\t%s\t%s\n", tag->TAGNAME, tag->TAGFILE, tag->TAGADDR);
		}
		else if (output_verbose_info)
		{
			for (i = 0; i < MAXATTR; i++)
			{
				if (tag->attr[i])
				{
					printf("%10s:%s\n",
						tagattrname[i], tag->attr[i]);
				}
			}
			printf("     match:%ld\n", tag->match);
			if (tag->next && output_all_matches)
				putchar('\n');
		}
		else if (output_html_browser)
		{
			printf("<tr>\n");
			printf("<th><a href=\"%s?", tag->TAGFILE);
			for (scan = tag->TAGADDR; *scan; scan++)
			{
				switch (*scan)
				{
				  case '\t':
				  case '+':
				  case '"':
				  case '%':
					printf("%%%02X", *scan);
					break;

				  case ' ':
					putchar('+');
					break;

				  default:
					putchar(*scan);
				}
			}
			printf("\">%s</a></th>\n", tag->TAGNAME);
			printf("<td>%s</td>\n", tag->TAGFILE);
			printf("<td>");
			if (elvdigit(*tag->TAGADDR))
			{
				printf("<em>line %s</em>", tag->TAGADDR);
			}
			else
			{
				for (scan = tag->TAGADDR + 2; scan[2]; scan++)
				{
					switch (*scan)
					{
					  case '&':  printf("&amp;");	break;
					  case '<':  printf("&lt;");	break;
					  case '>':  printf("&gt");	break;
					  case '"':  printf("&quot;");	break;
					  case '\\': scan++; /* and fall thru */
					  default:   putchar(*scan);
					}
				}
			}
			printf("</td>\n</tr>\n");
		}
		else
		{
			lookup(tag);
			if (tag->next && output_all_matches && !omit_comment_lines)
				puts("-------------------------------------------------------------------------------");
		}

		/* if supposed to stop after first, then stop */
		if (!output_all_matches)
			break;
	}

	/* output the HTML trailer, if appropriate */
	if (output_html_browser)
	{
		printf("</table>\n");
		printf("</body></html>");
	}

	/* done! */
	exit(0);
	return 0;	/* <- to silence a compiler warning */
}
